src/Microsoft.ML.TorchSharp/Roberta/Models/RobertaModel.cs

```diff
@@ -0,0 +1,133 @@
+        protected readonly LayerNorm LayerNorm;
+
+        protected readonly RobertaEncoder Encoder;
+
+        private const int PadIndex = 1;
+        private const int EosIndex = 2;
+        public override TransformerEncoder GetEncoder() => Encoder;
+#pragma warning restore MSML_PrivateFieldName // Private field name not in: _camelCase format
+
+        protected RobertaModel(QATrainer.Options options)
+            : base(options)
+        {
+            var negBillion = (int)-1e9;
+            Positions = Enumerable.Range(0, options.MaxSequenceLength).ToArray();
+            Zeros = Enumerable.Repeat(0, options.MaxSequenceLength).ToArray();
+            Ones = Enumerable.Repeat(1, options.MaxSequenceLength).ToArray();
+            NegBillionPad = Enumerable.Repeat(negBillion, options.MaxSequenceLength).ToArray();
+
+            Encoder = new RobertaEncoder(
+                numLayers: 12,
+                numAttentionHeads: 12,
+                numEmbeddings: 50265,
+                embeddingSize: 768,
+                hiddenSize: 768,
+                outputSize: 768,
+                ffnHiddenSize: 3072,
+                maxPositions: 512,
+                maxTokenTypes: 2,
+                layerNormEps: 1e-12,
+                embeddingDropoutRate: 0.1,
+                attentionDropoutRate: 0.1,
+                attentionOutputDropoutRate: 0.1,
+                outputDropoutRate: 0.1);
+        }
+
+        protected void InitWeights(torch.nn.Module module)
+        {
+            using var disposeScope = torch.NewDisposeScope();
+            if (module is Linear linearModule)
+            {
+                linearModule.weight.normal_(mean: 0.0, std: 0.02);
+                if (linearModule.bias.IsNotNull())
+                {
+                    linearModule.bias.zero_();
+                }
+            }
+            else if (module is Embedding embeddingModule)
+            {
+                embeddingModule.weight.normal_(mean: 0.0, std: 0.02);
+                embeddingModule.weight[1].zero_();  // padding_idx
+            }
+            else if (module is LayerNorm layerNormModule)
+            {
+                layerNormModule.weight.fill_(1.0);
+                layerNormModule.bias.zero_();
+            }
+        }
+
+        /// <summary>
+        /// Run only Encoder and return features.
+        /// </summary>
+        protected torch.Tensor ExtractFeatures(torch.Tensor srcTokens)
+        {
+            var (positions, segments, attentions) = GetEmbeddings(srcTokens);
+            var encodedVector = Encoder.call(srcTokens, positions, segments, attentions);
+            return encodedVector;
+        }
+
+        private (torch.Tensor position, torch.Tensor segment, torch.Tensor attentionMask) GetEmbeddings(torch.Tensor srcTokens)
+        {
+            using var disposeScope = torch.NewDisposeScope();
+            var device = srcTokens.device;
+            var positions = new torch.Tensor[srcTokens.size(0)];
```

Comment:
You call this method a lot.  Can you instead just store the result, or is it possible for srcTokens to change size during execution of this method?

Reply:
It doesn't change, caching it now.